go.mod

```diff
@@ -55,16 +55,37 @@ require (
 	k8s.io/klog v1.0.0
 	k8s.io/kube-aggregator v0.17.6
 	k8s.io/kube-openapi v0.0.0-20200410145947-bcb3869e6f29
+	k8s.io/kubernetes v1.17.6
```

Comment:
Then you may have to rewrite some code until package proxy is extracted to K8s staging projects.
Importing k8s.io/kubernetes directly will bring some burden when maintaining dependencies, it's highly discouraged. Projects that imported it are on the way dropping it. 
https://kubernetes.io/blog/2019/03/22/kubernetes-end-to-end-testing-for-everyone/#vendoring
https://github.com/helm/helm/issues/6606
https://github.com/kubernetes/minikube/issues/4656
https://github.com/vmware-tanzu/octant/pull/931

Reply:
Thanks for your information.
I copied some upstream codes to `pkg/agent/proxy/upstream`, here is the cloc information:
```
 Language             Files        Lines        Blank      Comment         Code
--------------------------------------------------------------------------------
 Go                      10         1337          178          366          793
```
I have cut all functions that are not directly used in our project.

Reply:
We need to be mindful about copying code. Is it the only option? If yes, make sure that we respect the terms of the Apache license below.

I would recommend doing the following:
* add a README.md file in that folder stating where the code comes from (link should include the k8s version), and that you removed parts of it
* add a comment to each file that you modified (see (b) below)
* even though we have the same license as K8s, I would recommend copying the Apache 2 LICENSE file again to that directory

It may be easier to copy the files without modifying them, this way we can ignore (b) and my second bullet point above.

```
   4. Redistribution. You may reproduce and distribute copies of the
      Work or Derivative Works thereof in any medium, with or without
      modifications, and in Source or Object form, provided that You
      meet the following conditions:

      (a) You must give any other recipients of the Work or
          Derivative Works a copy of this License; and

      (b) You must cause any modified files to carry prominent notices
          stating that You changed the files; and

      (c) You must retain, in the Source form of any Derivative Works
          that You distribute, all copyright, patent, trademark, and
          attribution notices from the Source form of the Work,
          excluding those notices that do not pertain to any part of
          the Derivative Works; and

      (d) If the Work includes a "NOTICE" text file as part of its
          distribution, then any Derivative Works that You distribute must
          include a readable copy of the attribution notices contained
          within such NOTICE file, excluding those notices that do not
          pertain to any part of the Derivative Works, in at least one
          of the following places: within a NOTICE text file distributed
          as part of the Derivative Works; within the Source form or
          documentation, if provided along with the Derivative Works; or,
          within a display generated by the Derivative Works, if and
          wherever such third-party notices normally appear. The contents
          of the NOTICE file are for informational purposes only and
          do not modify the License. You may add Your own attribution
          notices within Derivative Works that You distribute, alongside
          or as an addendum to the NOTICE text from the Work, provided
          that such additional attribution notices cannot be construed
          as modifying the License.

      You may add Your own copyright statement to Your modifications and
      may provide additional or different license terms and conditions
      for use, reproduction, or distribution of Your modifications, or
      for any such Derivative Works as a whole, provided Your use,
      reproduction, and distribution of the Work otherwise complies with
      the conditions stated in this License.
```

Reply:
Thanks, Antonin.
I have updated this PR according to your suggestion.

If we want to remove the dependency directly to Kubernetes we have to copy these files at least. I need to edit several codes also to avoid imports.